src/mewbot/api/type_guard.py

```diff
@@ -0,0 +1,97 @@
+# SPDX-FileCopyrightText: 2023 Mewbot Developers <mewbot@quicksilver.london>
+#
+# SPDX-License-Identifier: BSD-2-Clause
+
+"""
+Type guard exists to provide methods for run time typing validation.
+
+E.g. a decorator which checks that an InputEvent being passed to a function is of appropriate
+     type.
+
+The intent of the tools here is to provide a more elegant alternative to constructs like
+
+.. code-block:: python
+
+    def matches(self, event: InputEvent) -> bool:
+        if not isinstance(event, DiscordMessageCreationEvent):
+            return False
+        [Function body]
+
+which could be found in a Trigger.
+
+Instead, using one of the decorators provided in this module, you might write
+
+
+.. code-block:: python
+
+    @check_input_event_against_sig_types
+    def matches(self, event: InputEvent) -> bool:
+        [Function body]
+
+Or, as a more full example
+
+.. code-block:: python
+
+    # Demo Code
+
+    from mewbot.io.discord import DiscordInputEvent, DiscordUserJoinInputEvent
+    from mewbot.io.http import IncomingWebhookEvent
+
+    class A(Trigger):
+
+        @staticmethod
+        def consumes_inputs() -> set[type[InputEvent]]:
+            return {DiscordInputEvent, IncomingWebhookEvent}
+
+        @check_input_event_against_sig_types
+        def matches(self, event: DiscordInputEvent | IncomingWebhookEvent) -> bool:
+            return True
+
+    print(A.matches) # <function A.matches at 0x7f6e703f4a40>
+    print(A.matches.__doc__) # <function A.matches at 0x7f6e703f4a40>
+    print(A().matches(InputEvent())) # False
+    print(A().matches(DiscordInputEvent())) # True
+    print(A().matches(DiscordUserJoinInputEvent("[some text]"))) # True
+
+"""
+
+import functools
+
+from typing import Callable, TypeVar, get_type_hints, Union
+
+from mewbot.api.v1 import Trigger, Condition, InputEvent
+
+Component = TypeVar("Component", bound=Union[Trigger, Condition])
```

Comment:
If we're doing this 'properly', we should either move this into `mewbot.api.v1` (possible as a re-exported import from a sub-module), or change these to use the `*Interface` protocols from `mewbot.core`.

Reply:
(My inclination is to move it to `v1` because we might be able to have it applied automatically, so users can just declare their functions and stuff _works_)

Reply:
That makes sense to me - I'll move it into v1